contrib/babelfishpg_tsql/src/pltsql_bulkcopy.c

```diff
@@ -0,0 +1,938 @@
+				Datum *Values, bool *Nulls);
+
+static List *
+BulkCopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist);
+
+void BulkCopyErrorCallback(void *arg);
+
+/*
+ *	 BulkCopy - executes the Insert Bulk into a table
+ *
+ * Do not allow Bulk Copy if user doesn't have proper permission to access
+ * the table or the specifically requested columns.
+ */
+void
+BulkCopy(BulkCopyStmt *stmt, uint64 *processed)
+{
+	Relation	rel;
+	TupleDesc	tupDesc;
+	List	   *attnums;
+
+	Assert (stmt && stmt->relation);
+
+	/* Open and lock the relation, using the appropriate lock type. */
+	rel = table_openrv(stmt->relation, RowExclusiveLock);
+
+	tupDesc = RelationGetDescr(rel);
+
+	/* Generate or convert list of columns to process. */
+	attnums = BulkCopyGetAttnums(tupDesc, rel, stmt->attlist);
+
+	/* Execute Bulk Copy within try-catch block. */
+	PG_TRY();
+	{
+		if (!stmt->cstate)
+			stmt->cstate = BeginBulkCopy(rel, attnums);
+		
+		*processed = ExecuteBulkCopy(stmt->cstate, stmt->nrow, stmt->ncol, stmt->Values, stmt->Nulls);
+		stmt->rows_processed += *processed;
+	}
+	PG_CATCH();
+	{
+		/* For exact row which caused error, we have BulkCopyErrorCallback. */
+		elog(WARNING, "Error while executing Bulk Copy. Error occured while processing at "
+			"implicit Batch number: %d, Rows inserted in total: %ld", stmt->cur_batch_num, stmt->rows_processed);
+		if (rel != NULL)
+			table_close(rel, NoLock);
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	elog(DEBUG2, "Bulk Copy Progress: Successfully inserted implicit number of batches: %d, "
+		"number of rows inserted in total: %ld, "
+		"number of rows inserted in current batch: %ld",
+		stmt->cur_batch_num, stmt->rows_processed, *processed);
+
+	if (rel != NULL)
+		table_close(rel, NoLock);
+}
+
+
+/*
+ * BulkCopyGetAttnums - build an integer list of attnums to be copied
+ *
+ * The input attnamelist is either the user-specified column list,
+ * or NIL if there was none (in which case we want all the non-dropped
+ * columns).
+ *
+ * We don't include generated columns in the generated full list and we don't
+ * allow them to be specified explicitly.
+ *
+ * rel can be NULL ... it's only used for error reports.
+ */
+static List *
+BulkCopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist)
+{
+	List	   *attnums = NIL;
+
+	if (attnamelist == NIL)
+	{
+		/* Generate default column list. */
+		int			attr_count = tupDesc->natts;
+		int			i;
+
+		for (i = 0; i < attr_count; i++)
+		{
+			if (TupleDescAttr(tupDesc, i)->attisdropped)
+				continue;
+			if (TupleDescAttr(tupDesc, i)->attgenerated)
+				continue;
+			attnums = lappend_int(attnums, i + 1);
+		}
+	}
+	else
+	{
+		/* Validate the user-supplied list and extract attnums. */
+		ListCell   *l;
+
+		foreach(l, attnamelist)
+		{
+			char	   *name = (char *)lfirst(l);
+			int			attnum;
+			int			i;
+
+			/* Lookup column name. */
+			attnum = InvalidAttrNumber;
+			for (i = 0; i < tupDesc->natts; i++)
```

Comment:
This loop is looking up for the next attnum that matches with user provided column-name (for each column-name in column-list). Check line 204-205.
I believe It has to be done this way.

Reply:
you can go through each attribute and see if it matches one of provided column names. Not sure why it can't happen other way around

Reply:
It is similar, we required to have a separate function since the list attnamelist is list of "Name" structures which gets created by gram.y parser, whereas in ANTLR we get a List of strings. Moreover Rowversion datatype are allowed in BulkCopy.